# Programación Estructurada

In [9]:
def funcion(x, y):
    return x**2 * (4 - 2.1 * x**2 + x**4 / 3.0) + x * y + y**2 * (-4 + 4 * y**2)

# Búsqueda de Mínimos con Programación Estructurada
def busqueda(x_min, x_max, y_min, y_max, paso):
    mejor_x = x_min
    mejor_y = y_min
    mejor_valor = funcion(mejor_x, mejor_y)

    x = x_min
    while x <= x_max:
        y = y_min
        while y <= y_max:
            valor = funcion(x, y)

            if valor < mejor_valor:
                mejor_valor = valor
                mejor_x = x
                mejor_y = y

            y += paso
        x += paso

    return mejor_x, mejor_y, mejor_valor

# Parámetros de búsqueda
x_min = -10.0
x_max = 10.0
y_min = -10.0
y_max = 10.0
paso = 0.01

resultado = busqueda(x_min, x_max, y_min, y_max, paso)

print("Solución con programación estructurada:")
print("x =", resultado[0])
print("y =", resultado[1])
print("f(x,y) =", resultado[2])

Solución con programación estructurada:
x = 0.0899999999998311
y = -0.7100000000001694
f(x,y) = -1.031570363853008


# Algoritmo de enjambre de partículas

In [10]:
from random import random

def funcion(x, y):
    sum1 = x**2 * (4 - 2.1 * x**2 + x**4 / 3.0)
    sum2 = x * y
    sum3 = y**2 * (-4 + 4 * y**2)
    return sum1 + sum2 + sum3

def aleatorio(inf, sup):
    return random() * (sup - inf) + inf

class Particula:
    inercia = 1.4
    cognitiva = 2.0
    social = 2.0
    infx = -10.0
    supx = 10.0
    infy = -10.0
    supy = 10.0
    ajusteV = 100.0

    def __init__(self):
        self.x = aleatorio(Particula.infx, Particula.supx)
        self.y = aleatorio(Particula.infy, Particula.supy)
        self.vx = aleatorio(
            Particula.infx / Particula.ajusteV, Particula.supx / Particula.ajusteV
        )
        self.vy = aleatorio(
            Particula.infy / Particula.ajusteV, Particula.supy / Particula.ajusteV
        )
        self.xLoc = self.x
        self.yLoc = self.y
        self.valorLoc = funcion(self.x, self.y)

    def actualizaVelocidad(self, xGlob, yGlob):
        cogX = Particula.cognitiva * random() * (self.xLoc - self.x)
        socX = Particula.social * random() * (xGlob - self.x)
        self.vx = Particula.inercia * self.vx + cogX + socX
        cogY = Particula.cognitiva * random() * (self.yLoc - self.y)
        socY = Particula.social * random() * (yGlob - self.y)
        self.vy = Particula.inercia * self.vy + cogY + socY

    def actualizaPosicion(self):
        self.x = self.x + self.vx
        self.y = self.y + self.vy
        self.x = max(self.x, Particula.infx)
        self.x = min(self.x, Particula.supx)
        self.y = max(self.y, Particula.infy)
        self.y = min(self.y, Particula.supy)
        valor = funcion(self.x, self.y)
        if valor < self.valorLoc:
            self.xLoc = self.x
            self.yLoc = self.y
            self.valorLoc = valor


def enjambreParticulas(particulas, iteraciones, reduccionInercia):
    mejorParticula = min(particulas, key=lambda p: p.valorLoc)
    xGlob = mejorParticula.xLoc
    yGlob = mejorParticula.yLoc
    valorGlob = mejorParticula.valorLoc

    for iter in range(iteraciones):
        for p in particulas:
            p.actualizaVelocidad(xGlob, yGlob)
            p.actualizaPosicion()

        mejorParticula = min(particulas, key=lambda p: p.valorLoc)
        if mejorParticula.valorLoc < valorGlob:
            xGlob = mejorParticula.xLoc
            yGlob = mejorParticula.yLoc
            valorGlob = mejorParticula.valorLoc

        Particula.inercia *= reduccionInercia

    return (xGlob, yGlob, valorGlob)


nParticulas = 10
iteraciones = 100
redInercia = 0.9

particulas = [Particula() for i in range(nParticulas)]

print(enjambreParticulas(particulas, iteraciones, redInercia))

(-0.0898420184862, 0.712656402734921, -1.0316284534898772)


# Conclusiones

## Análisis de Comparación entre Algoritmos

Al comparar ambos algoritmos, se observa que los dos logran encontrar soluciones muy cercanas al mínimo global de la función, lo que demuestra que ambos métodos son adecuados para resolver el problema. La programación estructurada, mediante la búsqueda por barrido, encontró la solución en ((0.09, -0.71)) con un valor de (f(x,y) = -1.031570363853008), lo cual confirma que este método es confiable, ya que evalúa sistemáticamente muchos puntos dentro del intervalo y no depende del azar. Sin embargo, su precisión está limitada por el tamaño del paso utilizado y puede volverse más lenta al analizar grandes cantidades de combinaciones. Por otro lado, el algoritmo de enjambre de partículas encontró la solución en ((-0.08984201262945495, 0.7126564010127131)) con un valor de (f(x,y) = -1.0316284534898774), el cual es ligeramente menor, indicando una mejor aproximación al mínimo global. Esto se debe a que la Optimización por Enjambre de Partículas utiliza un enfoque más dinámico, donde las partículas exploran el espacio guiadas por su propia experiencia y la del grupo, permitiendo converger más eficientemente hacia soluciones óptimas. No obstante, al ser un método que utiliza aleatoriedad, sus resultados pueden variar en cada ejecución. En conclusión, la programación estructurada es más simple y determinista, mientras que el enjambre de partículas ofrece mayor eficiencia y precisión en la búsqueda de soluciones óptimas.


## Resultados Arrojados por Wolfram
Mínimos globales:

min{x^2 (4 - 2.1 x^2 + x^4/3) + x y + y^2 (-4 + 4 y^2)}≈-1.03163 en (x, y)≈(-0.089842, 0.712656)


min{x^2 (4 - 2.1 x^2 + x^4/3) + x y + y^2 (-4 + 4 y^2)}≈-1.03163 en (x, y)≈(0.089842, -0.712656)